wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java

```diff
@@ -506,6 +550,285 @@ private MethodSpec messageHashCode(MessageType type) {
+    result.beginControlFlow("if (serializedSize == -1)");
+
+    CodeBlock.Builder expression = CodeBlock.builder()
+        .add("message.unknownFieldsSerializedSize()");
+
+    if (!type.extensions().isEmpty()) {
+      expression.add("\n+ message.extensionsSerializedSize()");
+    }
+
+    for (Field field : type.fieldsAndOneOfFields()) {
+      expression.add("\n+ ");
+
+      int tag = field.tag();
+      String name = sanitize(field.name());
+      CodeBlock adapter = typeAdapter(field.type());
+      if (field.isRepeated() && !field.isPacked()) {
+        expression.add("sizeOfRepeated($L, message.$N, $L)", tag, name, adapter);
+      } else {
+        if (field.isPacked()) {
+          expression.add("sizeOfPacked($L, message.$N, $L)", tag, name, adapter);
+        } else {
+          expression.add("sizeOf($L, message.$N, $L)", tag, name, adapter);
+        }
+      }
+    }
+    result.addStatement("serializedSize = $L", expression.build());
+
+    result.endControlFlow();
+    result.addStatement("message.serializedSize = serializedSize");
+    result.addStatement("return serializedSize");
+    return result.build();
+  }
+
+  // Example:
+  //
+  // @Override
+  // public void write(SomeMessage message, ProtoWriter writer) {
+  //     if (opt_int32 != null) writer.value(1, message.opt_int32, TagWriter.INT32);
+  //     writer.value(2, message.req_int32, TagWriter.INT32);
+  //     message.writeUnknownFields(writer);
+  //   }
+  // }
+  //
+  private MethodSpec messageAdapterWrite(MessageType type, ClassName typeName) {
+    MethodSpec.Builder result = MethodSpec.methodBuilder("write")
+        .addAnnotation(Override.class)
+        .addModifiers(PUBLIC)
+        .returns(void.class)
+        .addParameter(typeName, "message")
+        .addParameter(ProtoWriter.class, "writer")
+        .addException(IOException.class);
+
+    for (Field field : type.fieldsAndOneOfFields()) {
+      int tag = field.tag();
+      String name = sanitize(field.name());
+      CodeBlock adapter = typeAdapter(field.type());
+      if (field.isRepeated()) {
+        if (field.isPacked()) {
+          result.addStatement("writer.packed($L, message.$N, $L)", tag, name, adapter);
+        } else {
+          result.addStatement("writer.repeated($L, message.$N, $L)", tag, name, adapter);
+        }
+      } else {
+        result.addStatement("writer.value($L, message.$N, $L)", tag, name, adapter);
+      }
+    }
+
+    if (!type.extensions().isEmpty()) {
+      result.addStatement("message.writeExtensions(writer)");
+    }
+
+    result.addStatement("message.writeUnknownFields(writer)");
+
+    return result.build();
+  }
+
+  // Example:
+  //
+  // @Override
+  // public Person read(ProtoReader reader) throws IOException {
+  //   Builder builder = new Builder();
+  //   while (reader.hasNext()) {
+  //     int tag = reader.nextTag();
+  //     switch (tag) {
+  //       case 1:
+  //         builder.name = reader.value(ProtoType.STRING);
```

Comment:
Then I need 30 methods, one for each scalar + message + enum and then repeated and packed variants of each. The visitor approach needs only 3, one for each arrangement of bytes.